### PR TITLE
[Samples][13.core-bot] Add BookingDialog and DateResolver classes

### PR DESCRIPTION
**Pull request by matiasroldan6**
Fixes #

## Specific Changes

  - 
  -
  -

### CODE REVIEW

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import com.microsoft.recognizers.datatypes.timex.expression.Constants;
+import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ *
+ */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:25]: I didn't see any doc in [C#](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/13.core-bot/Dialogs/BookingDialog.cs#L13), neither in [TS](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts#L22). We can remove this.

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+public class BookingDialog extends CancelAndHelpDialog {
+    private final String destinationStepMsgText = "Where would you like to travel to?";
+    private final String originStepMsgText = "Where are you traveling from?";
+
+    /**
+     *
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:32]: Delete it

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            this::confirmStep,
+            this::finalStep
+        };
+        addDialog(new WaterfallDialog(WaterfallDialog.class.getName(), Arrays.asList(waterfallSteps)));
+
+        // The initial child Dialog to run.
+        setInitialDialogId(WaterfallDialog.class.getName());
+    }
+
+    private CompletableFuture<DialogTurnResult> destinationStep(WaterfallStepContext stepContext) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:52]: Need doc, check [JS](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts#L40)

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                MessageFactory.text(destinationStepMsgText, destinationStepMsgText, InputHints.EXPECTING_INPUT);
+            return stepContext.prompt(TextPrompt.class.getName(), new PromptOptions() {{
+                setPrompt(promptMessage);
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getDestination());
+    }
+
+    private CompletableFuture<DialogTurnResult> originStep(WaterfallStepContext stepContext) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:66]: Need doc, check [JS](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts#L55)

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        bookingDetails.setDestination(stepContext.getResult().toString());
+
+        if (StringUtils.isNotBlank(bookingDetails.getOrigin())) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:71]:
```suggestion
        if (bookingDetails.getOrigin() == null) {
```

[C#](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/13.core-bot/Dialogs/BookingDialog.cs#L56) checks if it's null

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        bookingDetails.setTravelDate(stepContext.getResult().toString());
+
+        if (StringUtils.isNotBlank(bookingDetails.getTravelDate()) || isAmbiguous(bookingDetails.getTravelDate())) {
+            return stepContext.beginDialog(DateResolverDialog.class.getName(), bookingDetails.getTravelDate());
+        }
+
+        return stepContext.next(bookingDetails.getTravelDate());
+    }
+
+    private CompletableFuture<DialogTurnResult> confirmStep(WaterfallStepContext stepContext) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:94]: Need doc, check [JS](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts#L88)

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getOrigin());
+    }
+
+    private CompletableFuture<DialogTurnResult> travelDateStep(WaterfallStepContext stepContext) {
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        bookingDetails.setTravelDate(stepContext.getResult().toString());
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:85]:
```suggestion
        bookingDetails.setOrigin(stepContext.getResult().toString());
```

[C#](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/13.core-bot/Dialogs/BookingDialog.cs#L69) sets `origin`, not to `travelDate`

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+        if (StringUtils.isNotBlank(bookingDetails.getTravelDate()) || isAmbiguous(bookingDetails.getTravelDate())) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:87]:
```suggestion
        if (bookingDetails.getTravelDate() == null || isAmbiguous(bookingDetails.getTravelDate())) {
```

[C#](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/13.core-bot/Dialogs/BookingDialog.cs#L71) checks null

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                MessageFactory.text(originStepMsgText, originStepMsgText, InputHints.EXPECTING_INPUT);
+            return stepContext.prompt(TextPrompt.class.getName(), new PromptOptions() {{
+                setPrompt(promptMessage);
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getOrigin());
+    }
+
+    private CompletableFuture<DialogTurnResult> travelDateStep(WaterfallStepContext stepContext) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:82]: Need doc, check [JS](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts#L72)

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        return stepContext.next(bookingDetails.getTravelDate());
+    }
+
+    private CompletableFuture<DialogTurnResult> confirmStep(WaterfallStepContext stepContext) {
+        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
+
+        bookingDetails.setTravelDate(stepContext.getResult().toString());
+
+        String messageText =
+            String.format("Please confirm, I have you traveling to: %1s from: %2s on: %3s. Is this correct?",
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:100]:
```suggestion
            String.format("Please confirm, I have you traveling to: %s from: %s on: %s. Is this correct?",
```

If we use the parameters by order, it's not necessary to specify the index

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                bookingDetails.getDestination(), bookingDetails.getOrigin(), bookingDetails.getTravelDate());
+        Activity promptMessage = MessageFactory.text(messageText, messageText, InputHints.EXPECTING_INPUT);
+
+        return stepContext.prompt(ConfirmPrompt.class.getName(), new PromptOptions() {{
+            setPrompt(promptMessage);
+        }});
+    }
+
+    private CompletableFuture<DialogTurnResult> finalStep(WaterfallStepContext stepContext) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:109]: Need doc, check [JS](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts#L103)

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ *
+ */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:29]: Remove this

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+public class DateResolverDialog extends CancelAndHelpDialog {
+    private final String promptMsgText = "When would you like to travel?";
+    private final String repromptMsgText =
+        "I'm sorry, to make your booking please enter a full travel date including Day Month and Year.";
+
+    /**
+     *
+     * @param id
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:38]: Remove this

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        super(id != null ? id : DateResolverDialog.class.getName());
+
+        addDialog(new DateTimePrompt(DateTimePrompt.class.getName(), DateResolverDialog::dateTimePromptValidator));
+        WaterfallStep[] waterfallSteps = {
+            this::initialStep,
+            this::finalStep
+        };
+        addDialog(new WaterfallDialog(WaterfallDialog.class.getName(), Arrays.asList(waterfallSteps)));
+
+        setInitialDialogId(WaterfallDialog.class.getName());
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:49]:
```suggestion
        // The initial child Dialog to run.
        setInitialDialogId(WaterfallDialog.class.getName());
```

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    private CompletableFuture<DialogTurnResult> initialStep(WaterfallStepContext stepContext) {
+        String timex = stepContext.getOptions().toString();
+
+        Activity promptMessage = MessageFactory.text(promptMsgText, promptMsgText, InputHints.EXPECTING_INPUT);
+        Activity repromptMessage = MessageFactory.text(repromptMsgText, repromptMsgText, InputHints.EXPECTING_INPUT);
+
+        if (StringUtils.isBlank(timex)) {
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:58]:
```suggestion
        if (timex == null) {
```
[C#](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/13.core-bot/Dialogs/DateResolverDialog.cs#L40) checks for null

**Review by Batta32 — CHANGES_REQUESTED** [review]
Check JS to add the comments, use it as a guide:
* [bookingDialog](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/bookingDialog.ts)
* [dateResolverDialog](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/13.core-bot/src/dialogs/dateResolverDialog.ts)

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java (hunk 1 above)
Comment (matiasroldan6) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:25]: This is required by Checkstyle, which demands a Javadoc documentation of this. Deleting it creates an error. I expected to complete the full doc in the fixing errors stage

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java (hunk 2 above)
Comment (matiasroldan6) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:32]: same as above

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java (hunk 13 above)
Comment (matiasroldan6) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:38]: all Javadocs were required by Checkstyle

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java (hunk 7 above)
Comment (matiasroldan6) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:85]: my bad

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java (hunk 5 above)
Comment (matiasroldan6) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:71]: I've been looking at these comments. I've doubts -at least in [C#](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/13.core-bot/Dialogs/BookingDialog.cs#L56)- that string can ever be null, unless it's a class parsed and properties are generated. I'll change it to null anyway in our Java code, and we can later check the bot running

Comment (Batta32) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:71]: Ack, however, you added `isNotBlank` and it should be `isBlank` in this case. `isBlank` check if it's null or an empty String

Comment (matiasroldan6) [reply samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java]: yes, the check was reversed, I realized now

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+import com.microsoft.bot.schema.InputHints;
+import com.microsoft.recognizers.datatypes.timex.expression.Constants;
+import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;
+
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ *
+ */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:24]: As we won't remove this, try to fill them

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+public class BookingDialog extends CancelAndHelpDialog {
+    private final String destinationStepMsgText = "Where would you like to travel to?";
+    private final String originStepMsgText = "Where are you traveling from?";
+
+    /**
+     *
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:31]: As we won't remove this, try to fill them

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+                setPrompt(promptMessage);
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getDestination());
+    }
+
+    /**
+     * If an origin city has not been provided, prompt for one.
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:70]: We are missing the `@param` and `@return`

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+            }});
+        }
+
+        return stepContext.next(bookingDetails.getOrigin());
+    }
+
+    /**
+     * If a travel date has not been provided, prompt for one.
+     * This will use the DATE_RESOLVER_DIALOG.
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:90]: We are missing the `@param` and `@return`

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        if (bookingDetails.getTravelDate() == null || isAmbiguous(bookingDetails.getTravelDate())) {
+            return stepContext.beginDialog(DateResolverDialog.class.getName(), bookingDetails.getTravelDate());
+        }
+
+        return stepContext.next(bookingDetails.getTravelDate());
+    }
+
+    /**
+     * Confirm the information the user has provided.
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/BookingDialog.java:105]: We are missing the `@param` and `@return`

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ *
+ */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:28]: As we won't remove this, try to fill them

File: samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+public class DateResolverDialog extends CancelAndHelpDialog {
+    private final String promptMsgText = "When would you like to travel?";
+    private final String repromptMsgText =
+        "I'm sorry, to make your booking please enter a full travel date including Day Month and Year.";
+
+    /**
+     *
+     * @param id
+     */
```
Comment (Batta32) [inline samples/13.core-bot/src/main/java/com/microsoft/bot/sample/core/DateResolverDialog.java:37]: As we won't remove this, try to fill them

**Review by Batta32 — CHANGES_REQUESTED** [review]
Populate the java docs

**Review by Batta32 — APPROVED** [review]
LGTM

**Pull request merged**